### PR TITLE
feat(azure): support custom configuration

### PR DESCRIPTION
**Pull request by NicholasDawson**
### 🔗 Linked issue

#1087 

### ❓ Type of change

- [X] 📖 Documentation (updates to the documentation or readme)
- [ ] 🐞 Bug fix (a non-breaking change that fixes an issue)
- [X] 👌 Enhancement (improving an existing functionality like performance)
- [X] ✨ New feature (a non-breaking change that adds functionality)
- [ ] ⚠️ Breaking change (fix or feature that would cause existing functionality to change)

### 📚 Description

I, along with many others deploying a Nitro server to Azure Static Web Apps, need to customize the `staticwebapp.config.json` file beyond what Nitro generates. Currently, Nitro overwrites the entire file on every compilation. It would be better if Nitro extended an existing configuration if present not overwrite the whole file.

I added thorough documentation to explain how the system will now load and utilize the configuration. Please reference that to learn a little more about my implementation.

#### Implementation Notes
I put a lot of thought into this implementation as there wasn't necessarily a right or wrong way to handle this. I didn't originally want to have a separate file `custom.staticwebapp.config.json`, but the reason I had to add it was that if Nitro and the user are writing to the same file there is no way to keep track of what needs to take priority or be overwritten. Especially for routes, if Nitro generated the config for pre-rendered routes, but then the user deletes that pre-rendered route, it would be expected that the config removes that next time it compiles. If I tried to merge changes into the `staticwebapp.config.json` then Nitro would essentially lose the ability to remove lines from the configuration, only add them.

Therefore it was necessary to introduce the convention of having the user add their own configuration to a file I called `custom.staticwebapp.config.json` to provide a clear distinction between what the user wants to add and what the final generated result (`staticwebapp.config.json`) is.

### 📝 Checklist

- [X] I have linked an issue or discussion.
- [X] I have updated the documentation accordingly.

### CODE REVIEW

**Comment by NicholasDawson** [comment]
Can I please have an update on this PR? It has been 3 weeks, I tried really hard to add sufficient testing and documentation to make this PR easy to review and accept. I know you all must be very busy and I really appreciate your work, I would just appreciate an update on this PR.

**Comment by Hebilicious** [comment]
> @Hebilicious Can I please have an update on this PR? It has been 3 weeks, I tried really hard to add sufficient testing and documentation to make this PR easy to review and accept. I know you all must be very busy and I really appreciate your work, I would just appreciate an update on this PR.

Hi @NicholasDawson , thank you for your hard work on this PR. I think these are all great changes that are well documented, and they will make the life of azure users easier.
@pi0 will need to review it and approve it before it gets merged.

File: docs/content/2.deploy/providers/azure.md
Comment (Hebilicious) [inline docs/content/2.deploy/providers/azure.md:42]: Can be simplified : 
You can alter the Nitro generated configuration with a `custom.staticwebapp.config.json` file in your project root.

File: docs/content/2.deploy/providers/azure.md
Comment (Hebilicious) [inline docs/content/2.deploy/providers/azure.md:44]: We can simplify this : 

Custom routes will be added and matched first. In the case of a conflict (determined if an object has the same `route` property), custom routes will override generated ones.

File: src/presets/azure.ts
Comment (Hebilicious) [inline src/presets/azure.ts:47]: This can fail silently if JSON.parse fails (invalid json). 
Maybe we should actually throw an error if JSON.parse fails ? 
One way to do it would be to check if the file exist first.
 
`const customConfig = exists ? JSON.parse() : {}`

File: src/presets/azure.ts
Comment (Hebilicious) [inline src/presets/azure.ts:124]: could use `config.routes.findIndex` instead of some to simplify the logic here.

File: test/fixture/custom.staticwebapp.config.json
Comment (Hebilicious) [inline test/fixture/custom.staticwebapp.config.json:1]: Why do we need this file if we attempt to delete it when the tests runs ?

File: src/presets/azure.ts
Comment (Hebilicious) [inline src/presets/azure.ts:131]:
```suggestion
      if (existingRouteMatchIndex) {
```

**Comment by NicholasDawson** [comment]
> I would get rid of the try catch block entirely. Otherwise this looks good to me.

@Hebilicious All set.

**Comment by NicholasDawson** [comment]
@Hebilicious do you know if this PR is ready to be merged, or does @pi0 need to review as well? Just wondering where it stands.

**Comment by Hebilicious** [comment]
Pi is the code owner and has to review and merge all non documentation PRs.

**Review by pi0 — APPROVED** [review]
Thanks so much for the PR ❤️  (and sorry it took long to review!)

I have made a couple of reactors to support via `nitro.azure.config` instead to avoid introducing second file and be consistent with other presets i hope you are happy with it.

**Pull request merged**
